### DISCUSSION TITLE
Two quick Qs: Any way to tie eglot to a language hook, and popup.el support?

### DISCUSSION THREAD

**Discussion by ahungry** [joaotavora/eglot#1095] [Q&A]
First question is pretty simple and self explanatory - the second one, I'll try to explain a bit more.

As a non-GUI user, when I used flycheck + lsp, if an error was detected in buffer, I could move the cursor on top of the error (either with prev-error or next-error (sidenote: these seem like general functions of Emacs - why do I need a specific flymake-next-error or flymake-prev-error now?  it means I need contextual/mode specific keybinds if I want to retain my habitual bindings) and it would show the error message (the same one `C-h .` would show) at point.  Coincidentally, `C-h .` for `display-local-help` seems overshadowed by Eglot, so I can't use the default binding (it opens an entire side buffer with documentation when I perform it on a function that is grayed out due to no function calls/references).

If someone needs more context for the flycheck popup at point support in TTY mode https://github.com/flycheck/flycheck-popup-tip has a nice screenshot on the README page of what this looks like.

Is there anyway (short of me hacking it up myself) to provide this?  Is the lack of support/attention simply me missing a setting, or a problem with popup.el being a non-GNU/Emacs package, and there is no such native support for non-GUI users?

**Comment by ahungry** [comment]
Neat, I found https://github.com/meqif/flymake-diagnostic-at-point which seems to scratch my itch for a popup at point - now I'm just curious about hooking eglot - adding #'eglot as a mode hook doesn't seem to work properly...

**Reply by ssl19** [reply]
>  now I'm just curious about hooking eglot - adding #'eglot as a mode hook doesn't seem to work properly...

M-x `eglot`  when you open a project for the first time, then eglot will auto activate itself when open other files in this project. Check this: 
https://github.com/joaotavora/eglot/blob/e501275e06952889056268dabe08ccd0dbaf23e5/eglot.el#L1788

or add `eglot-ensure` to some-mode-hook (not encouraged).
https://github.com/joaotavora/eglot/pull/989#issuecomment-1173845943 
> I kind of regret having added eglot-ensure, but users seem to love it 🤷. Is remembering/typing M-x eglot when starting to work on a project really that much work?

**Reply by ahungry** [reply]
Thanks @ssl19 !  Remembering to run `M-x eglot` isn't much work, but it compounds with other things - for instance (ignoring the overlap/redundancy between the two) - if I open a Clojure project (of which I have many), I now have to:

1. Use project-switch-project to list my projects
2. Choose the main entry file I want to work on (core.clj most the time)
3. `M-x cider-jack-in-clj` to begin firing up CIDER
4. `M-x eglot` to begin the language server client

If I can instead simplify everything to "pick a project and it all starts" - that's like the very reason we use mode hooks for other features (why have a flymake/flycheck hook to turn it on?  Can't users just `M-x flymake-mode`?  What about the 10+ other minor modes)?

In an editor that encourages and supports users to enable power user workflows, typing *more* things manually/by hand is almost always less preferable than automation.

It's a really curious stance (is the disapproval due to some objective reason?  I do see eglot starts quite a bit slower than lsp-mode)

**Reply by ssl19** [reply]
> If I can instead simplify everything to "pick a project and it all starts" - that's like the very reason we use mode hooks for other features (why have a flymake/flycheck hook to turn it on? Can't users just `M-x flymake-mode`? What about the 10+ other minor modes)?

Then you can use `eglot-ensure`, according to  https://joaotavora.github.io/eglot/

The function eglot-ensure will start an Eglot session for each buffer in which foo-mode is turned on, if there isn’t already an Eglot session that handles the buffer. Note that this variant of starting an Eglot session is non-interactive, so it should be _**used only when you are confident**_ that Eglot can be started reliably for any file which may be visited with the major-mode in question. 

> It's a really curious stance (is the disapproval due to some objective reason? I do see eglot starts quite a bit slower than lsp-mode)

try `(setq eglot-events-buffer-size 0)` before starting the lsp server.
